prov/sm2/src/sm2.h

```diff
@@ -78,6 +79,11 @@
 #define SM2_IOV_LIMIT		4
 #define SM2_INJECT_SIZE		(SM2_XFER_ENTRY_SIZE - sizeof(struct sm2_xfer_hdr))
 
+#define SM2_ATOMIC_INJECT_SIZE \
+	(SM2_INJECT_SIZE - sizeof(struct sm2_atomic_secondary_hdr))
+#define SM2_ATOMIC_COMP_INJECT_SIZE (SM2_ATOMIC_INJECT_SIZE / 2)
```

Comment:
I was doing something similar for RMA.  Both RMA and Atomic have similar wording in the docs here:

> The requested message size that can be used with fi_inject_atomic is limited by inject_size.

I think that means that fi_inject_atomic, fi_inject_write, and fi_inject all must have the same limit, so we may need to report the minimum of these.

Reply:
I thought user got size through these functions:
```
static int sm2_atomic_fetch_valid(struct fid_ep *ep, enum fi_datatype datatype,
				  enum fi_op op, size_t *count);

...

	.writevalid = sm2_atomic_valid,
	.readwritevalid = sm2_atomic_fetch_valid,
	.compwritevalid = sm2_atomic_comp_valid,
```

Reply:
Users should definitely query the intended operation/datatype to figure out how many items they can call it on but that doesn't indicate anything about inject. The inject atomic/RMA whether you should pick the smallest value is kind of a grey zone... The way we get around it in shm is that inject size is the full payload buffer so doesn't depend on the headers and since an atomic inject is never a comp, we can use the full inject size so they can be the same. If you're planning on moving to a model similar to that (where the headers are separate than the inject buffer itself) then I think it is a nonissue. What is the plan with the buffer contents?

Reply:
I don't want to add anything else to the primary header, we should aim for the primary header to be as small as possible. I might even want to take context and user data out of the primary header and make them secondary header attachments if we find that it is faster. I don't mind casting the user data to be the start of the secondary header object. Still undecided on a small message size optimization (separate freestack vs header only freestack).

Reply:
hmm, then I think you're going to want to reduce the inject size to the minimum like Luke says, especially if you're thinking of including other, non-atomic things and including the header here. I think putting the second header in front of the inject buffer is not intuitive, weird, and not worth it. If you're really seeing a performance improvement with that vs leaving the stuff in the header and doing the full 4k for the inject then we can talk...

Reply:
Updated Inject size